src/handlers/csv-extract/parsing-utils/parse-csv.ts

```diff
@@ -0,0 +1,93 @@
+export interface InvoiceData {
+  vendor: string;
+  invoiceDate: string;
+  dueDate: string;
+  vatNumber: string;
+}
+
+const isKvpRow = (row: string[]): boolean => {
+  return (
+    !!row[0] &&
+    !!row[1] &&
+    row.every((cell, i) => {
+      if (i <= 1) return true;
+      return !cell;
+    })
+  );
+};
+
+const isFullRow = (row: string[]): boolean => {
+  return (
+    row !== undefined &&
+    row.length > 0 &&
+    row.every((cell) => {
+      return cell;
+    })
+  );
+};
+
+const isTotalRow = (row: string[]): boolean => {
+  return row[0] === "Total";
+};
+
+const objectifyLineItemRow = (
+  tableHeader: string[],
+  row: string[]
+): Record<string, string> => {
+  return tableHeader.reduce(
+    (acc, heading, j) => ({
+      ...acc,
+      [heading]: row[j],
+    }),
+    {}
+  );
+};
+
+const applyStructure = (
+  cells: string[][]
+): {
+  kvps: Record<string, string>;
+  lineItems: Array<Record<string, string>>;
+} => {
+  let tableHeader: string[] | null = null;
+  return cells.reduce(
+    (acc, row) => {
+      if (isKvpRow(row)) {
+        return {
+          ...acc,
+          kvps: { ...acc.kvps, [row[0]]: row[1] },
+        };
+      }
+      if (isFullRow(row) && !tableHeader) {
+        tableHeader = row;
+        return acc;
+      }
+      if (isTotalRow(row)) {
+        return acc;
+      }
+      if (isFullRow(row)) {
+        if (!tableHeader) {
+          throw new Error("Encountered a line item before finding a header.");
```

Comment:
I agree it shouldn’t ever really happen, I was probably being overly cautious, or maybe getting around a type error, I can’t remember

Reply:
If I get rid of this check, the next line is flagged as an error in my IDE at least because it's not smart enough to know that tableHeader isn't null.  So I've moved the logic around a bit in my fix.